cylc/flow/task_pool.py

```diff
@@ -2047,30 +2061,119 @@ def _get_active_flow_nums(self) -> Set[int]:
+            else:
+                itask.flow_nums.difference_update(fnums_to_remove)
+
+        matched_task_ids = {
+            *removed.keys(),
+            *(quick_relative_id(cycle, task) for task, cycle in inactive),
+        }
+
+        for id_ in matched_task_ids:
+            point_str, name = id_.split('/', 1)
+            tdef = self.config.taskdefs[name]
+            # Unset any prereqs naturally satisfied by these tasks
+            # (do not unset those satisfied by `cylc set --pre`):
+            for child in itertools.chain.from_iterable(
+                generate_graph_children(tdef, get_point(point_str)).values()
             ):
-                self.xtrigger_mgr.sequential_has_spawned_next.add(
-                    itask.identity
-                )
-                self.spawn_to_rh_limit(
-                    itask.tdef,
-                    itask.tdef.next_point(itask.point),
-                    itask.flow_nums
+                child_itask = self.get_task(child.point, child.name)
+                if not child_itask:
+                    continue
+                fnums_to_remove = child_itask.match_flows(flow_nums)
+                if not fnums_to_remove:
+                    continue
+                prereq_changed = False
+                for prereq in (
+                    *child_itask.state.prerequisites,
+                    *child_itask.state.suicide_prerequisites,
+                ):
+                    for msg in prereq.naturally_satisfied_dependencies():
+                        if msg.get_id() == id_:
+                            prereq[msg] = False
+                            prereq_changed = True
+                            if id_ not in removed:
+                                removed[id_] = fnums_to_remove
+                if prereq_changed:
+                    self.unqueue_task(child_itask)
+                    self.data_store_mgr.delta_task_prerequisite(child_itask)
+
+            # Remove from DB tables:
+            db_removed_fnums = self.workflow_db_mgr.remove_task_from_flows(
+                point_str, name, flow_nums
+            )
+            if db_removed_fnums:
+                removed.setdefault(id_, set()).update(db_removed_fnums)
+
+        if removed:
+            tasks_str_list = []
+            for task, fnums in removed.items():
+                self.data_store_mgr.delta_remove_task_flow_nums(task, fnums)
+                tasks_str_list.append(
+                    f"{task} {repr_flow_nums(fnums, full=True)}"
                 )
-            self.remove(itask, 'request')
-        if self.compute_runahead():
+            LOG.info(f"Removed task(s): {', '.join(sorted(tasks_str_list))}")
+
+        not_removed.update(matched_task_ids.difference(removed))
+        if not_removed:
+            fnums_str = (
+                repr_flow_nums(flow_nums, full=True) if flow_nums else ''
+            )
+            LOG.warning(
+                "Task(s) not removable: "
```

Comment:
I think that an explanation here might be helpful, perhaps:

> Task(s) does not exist in flow(s) so not removed:

Reply:
This message will be logged if the user tries to remove a task that either:
- has no intersecting flow numbers with what was specified (as you've said)
- has not run yet
- has already been removed

Should we try and explain this, or stick to the simple "not removable"?

Reply:
Honestly I'd prefer the longer explanation, but I agree, the alternative is a bit long. Can you not differentiate between the three causes?

Reply:
Nope

Reply:
That's fine. I don't believe this is a blocker.